Language/Haskell/GhcMod/Utils.hs

```diff
@@ -156,3 +160,41 @@ canonFilePath f = do
   e <- doesFileExist p
   when (not e) $ error $ "canonFilePath: not a file: " ++ p
   return p
+
+withMappedFile :: (IOish m, GmState m, GmEnv m) =>
+                  forall a. FilePath -> (FilePath -> m a) -> m a
+withMappedFile file action = getCanonicalFileNameSafe file >>= lookupMMappedFile >>= runWithFile
+  where
+    runWithFile (Just (RedirectedMapping to)) = action to
+    runWithFile (Just (MemoryMapping (Just src))) = do
+      crdl <- cradle
+      (fp,hndl) <- liftIO $ openTempFile (cradleTempDir crdl) (takeFileName file)
+      liftIO $ hPutStr hndl src
+      liftIO $ hClose hndl
+      result <- action fp
+      liftIO $ removeFile fp
+      return result
+    runWithFile _ = action file
+
+getCanonicalFileNameSafe :: (IOish m, GmEnv m) => FilePath -> m FilePath
```

Comment:
It's an arbitrary limitation with no convincing reason, so not a fan of this particular idea.

In all honesty, for my intended use-case, possibility of changing filesystem shouldn't be a problem, since I actually 'unmap' files once I'm done with them, so unless filesystem changes right in the middle of a command (which is not very likely), we're good. But I can construct a test case where this might actually fail (that would need symlinks though -- so not on Windows).

I think I might be able to defer canonicalization until file is actually used, which would somewhat mitigate this issue.